### PR TITLE
Up Arrow key cycles through Query history

### PR DESCRIPTION
**Pull request by Garulf**
Similar to how arrow keys work in most shells this PR allows the user to press <kbd>Up</kbd> to cycle through their last used queries.

# Todo:
- [x] Cycle through entire history if user hasn't moved selection
- [x] Fix not being able to move selection in History view
- [ ]  ~~Add custom hotkey support~~

https://user-images.githubusercontent.com/535299/144291288-73ffa191-5deb-4f42-9675-90b748a82e32.mp4

### CODE REVIEW

**Comment by onesounds** [comment]
- The same problem exists in the context menu.
- I think there should be a shortcut to erasing it.  (ctrl+u?)

**Comment by jjw24** [comment]
i cant seem to cycle past the first history that showed, basically if i keep pressing up after the first history it just cycle through the results from it.

also need to check and not trigger when in context menu

is this pr still draft?

**Comment by Garulf** [comment]
> i cant seem to cycle past the first history that showed, basically if i keep pressing up after the first history it just cycle through the results from it.
> 
> also need to check and not trigger when in context menu
> 
> is this pr still draft?

Yeah it's still just a proof-of-concept...

**Comment by Garulf** [comment]
https://user-images.githubusercontent.com/535299/145690824-1d2a8f17-d8ee-4463-a792-55eff6351b41.mp4

Working

**Comment by jjw24** [comment]
i am getting this error. repro:
1. only one history result present
2. up arrow to insert the query
3. alt + down arrow to forward search

image

File: Flow.Launcher/ViewModel/MainViewModel.cs [hunk 1]
```diff
@@ -29,16 +29,19 @@ public class MainViewModel : BaseModel, ISavable
         #region Private Fields
 
         private const string DefaultOpenResultModifiers = "Alt";
+        private const string DefaultCycleHistoryModifiers = "Alt";
 
         private bool _isQueryRunning;
         private Query _lastQuery;
         private string _queryTextBeforeLeaveResults;
 
+
```
Comment (jjw24) [inline Flow.Launcher/ViewModel/MainViewModel.cs:38]:
```suggestion
```

File: Flow.Launcher/ViewModel/MainViewModel.cs [hunk 2]
```diff
@@ -775,7 +833,7 @@ public async void Hide()
                 default:
                     throw new ArgumentException($"wrong LastQueryMode: <{_settings.LastQueryMode}>");
             }
-
+            _lastHistoryIndex = 1;
```
Comment (jjw24) [inline Flow.Launcher/ViewModel/MainViewModel.cs:836]: why is this needed here and why 1?

File: Flow.Launcher/ViewModel/MainViewModel.cs [hunk 3]
```diff
@@ -29,16 +29,19 @@ public class MainViewModel : BaseModel, ISavable
...
         private Query _lastQuery;
         private string _queryTextBeforeLeaveResults;
 
+
         private readonly FlowLauncherJsonStorage<History> _historyItemsStorage;
         private readonly FlowLauncherJsonStorage<UserSelectedRecord> _userSelectedRecordStorage;
         private readonly FlowLauncherJsonStorage<TopMostRecord> _topMostRecordStorage;
         internal readonly Settings _settings;
         private readonly History _history;
+        private int _lastHistoryIndex = 1;
```
Comment (jjw24) [inline Flow.Launcher/ViewModel/MainViewModel.cs:44]: this doesnt need to be underscore, code base may have some private variables with underscores but new code should use just lowercase and we will gradually rename these underscores. the standard naming practice for private variables is no longer underscore.

File: Flow.Launcher/ViewModel/MainViewModel.cs [hunk 4]
```diff
@@ -189,6 +193,51 @@ private void InitializeKeyCommands()
...
+            InsertLastQuery = new RelayCommand(_ => {
+                var results = SelectedResults;
+
+                if (_history.Items.Count > 0 
+                    && _queryText == String.Empty
+                    && !HistorySelected() 
+                    && !ContextMenuSelected())
+                {
+                    ChangeQueryText(_history.Items[_history.Items.Count - 1].Query.ToString());
+                    _lastHistoryIndex = 2;
```
Comment (jjw24) [inline Flow.Launcher/ViewModel/MainViewModel.cs:232]: why 2? this is the cause for the above error in the comment. try not to use plain numbers and strings please

**Review by jjw24 — CHANGES_REQUESTED** [review]
hey just a couple changes needed please :)

File: Flow.Launcher/ViewModel/MainViewModel.cs [hunk 5]
```diff
@@ -409,8 +458,12 @@ private ResultsViewModel SelectedResults
         public ICommand ReloadPluginDataCommand { get; set; }
         public ICommand ClearQueryCommand { get; private set; }
         public ICommand AutocompleteQueryCommand { get; set; }
+        public ICommand ReverseHistory { get; set; }
+        public ICommand ForwardHistory { get; set; }
+        public ICommand InsertLastQuery { get; set; }
```
Comment (jjw24) [inline Flow.Launcher/ViewModel/MainViewModel.cs:463]: these command variable naming should follow the others- e.g. ReverseHistoryCommand

**Comment by jjw24** [comment]
so i am finding it a bit weird when i press up arrow on an empty query and only the one history result (the last one shows) then the subsequent up arrow presses will just cycle through the result. what are your throughts, up arrow only is able to cycle up one result, do we want to do this where the alt + up/down arrows work like a treat.

**Comment by Garulf** [comment]
> so i am finding it a bit weird when i press up arrow on an empty query and only the one history result (the last one shows) then the subsequent up arrow presses will just cycle through the results. what are your throughts, up arrow only is able to cycle up one result, do we want to do this where the alt + up/down arrows work like a treat.

This intended behavior as the previous way got in the way of default behavior. Sorry I haven't documented yet as I was still testing.

File: Flow.Launcher/ViewModel/MainViewModel.cs (hunk 4 above)
Comment (Garulf) [reply Flow.Launcher/ViewModel/MainViewModel.cs:231]: Oh this is set so if the user hits <kbd>Up</kbd> on an empty Query and decides to use <kbd>Alt</kbd>+<kbd>Up</kbd> they don't repeat inserting the first item.

File: Flow.Launcher/ViewModel/MainViewModel.cs (hunk 2 above)
Comment (Garulf) [reply Flow.Launcher/ViewModel/MainViewModel.cs]: Resets the index when Flow is deactivated/hidden

File: Flow.Launcher/ViewModel/MainViewModel.cs (hunk 4 above)
Comment (Garulf) [reply Flow.Launcher/ViewModel/MainViewModel.cs]: I should add logic to make sure we dont set this to an index that doesn't exist

**Comment by Garulf** [comment]
Currently <kbd>Alt</kbd>+<kbd>Up</kbd> & <kbd>Alt</kbd>+<kbd>Down</kbd> Only trigger if defined in `MainWindow.xaml` and will not trigger from `MainWindow.xaml.cs`

File: Flow.Launcher/ViewModel/MainViewModel.cs (hunk 4 above)
Comment (Garulf) [reply Flow.Launcher/ViewModel/MainViewModel.cs]: Fixed: https://github.com/Flow-Launcher/Flow.Launcher/pull/852/commits/c6d5525dec46631db6cf587573d6ecc55c8b9419

**Comment by Garulf** [comment]
> i am getting this error. repro:
> 
> 1. only one history result present
> 2. up arrow to insert the query
> 3. alt + down arrow to forward search
> 
> image

Fixed: https://github.com/Flow-Launcher/Flow.Launcher/pull/852/commits/c6d5525dec46631db6cf587573d6ecc55c8b9419

**Comment by jjw24** [comment]
@Garulf this is ready for review again? (Merge Dev if so)

**Comment by Garulf** [comment]
Could not find a proper method to set <kbd>Alt</kbd> & <kbd>Ctrl</kbd> modifiers. I could gut it and hard code to <kbd>Alt</kbd> or keep as is?

**Comment by jjw24** [comment]
The modifier hotkey can set Alt, does that part of the code offer any help?

image

**Comment by Garulf** [comment]
> The modifier hotkey can set Alt, does that part of the code offer any help?
> 
> image

The problem is the current implementation requires the hotkey be coded in two separate files. Doesn't feel like a great fix.

**Comment by Garulf** [comment]
Removed from GUI for now

**Comment by taooceros** [comment]
close in favor of #2696

**Pull request closed without merging**